runtime/src/commonJvmAndroid/kotlin/pbandk/internal/binary/BinaryMessageDecoderJvm.kt

```diff
@@ -1,18 +1,35 @@
 package pbandk.internal.binary
 
-import com.google.protobuf.CodedInputStream
-import pbandk.*
+import pbandk.MessageDecoder
+import pbandk.internal.binary.kotlin.ByteArrayWireReader
+import pbandk.internal.binary.kotlin.KotlinBinaryWireDecoder
 import java.io.InputStream
 import java.nio.ByteBuffer
 
-internal actual fun BinaryMessageDecoder.Companion.fromByteArray(arr: ByteArray): MessageDecoder {
-    return BinaryMessageDecoder(CodedStreamBinaryWireDecoder(CodedInputStream.newInstance(arr)))
-}
+private fun fromByteArray(arr: ByteArray, offset: Int, length: Int) =
+    BinaryMessageDecoder(KotlinBinaryWireDecoder(ByteArrayWireReader(arr, offset, length)))
+
+// TODO: Maybe allow the caller to pass in optional offset and length for reading from the array
+internal actual fun BinaryMessageDecoder.Companion.fromByteArray(arr: ByteArray): MessageDecoder =
+    fromByteArray(arr, 0, arr.size)
 
-internal fun BinaryMessageDecoder.Companion.fromInputStream(stream: InputStream): MessageDecoder {
-    return BinaryMessageDecoder(CodedStreamBinaryWireDecoder(CodedInputStream.newInstance(stream)))
+// TODO: Maybe expose the [bufferSize] parameter from [InputStreamWireReader] to the caller
+internal fun BinaryMessageDecoder.Companion.fromInputStream(stream: InputStream, expectedSize: Int = -1): MessageDecoder {
+    val wireReader = if (expectedSize != -1) {
+        InputStreamWireReader(stream, expectedSize)
+    } else {
+        InputStreamWireReader(stream)
+    }
+    return BinaryMessageDecoder(KotlinBinaryWireDecoder(wireReader))
```

Comment:
I intentionally did not use `Int.MAX_VALUE` because I was concerned it artificially limits us to reading messages less than 2GB in size. While you'd probably be crazy to try to parse a 2GB message, it also seemed like something we might want to do in the future and it would be easy to support by taking a `Long` instead of an `Int` here. `-1` seemed like a safer choice that would never be passed in as a size value. Admittedly, this is optimizing for a very far away future, but it took so little to support this case that I figured it was worthwhile to do.

The reason `InputStreamWireReader` still takes `Int.MAX_VALUE` is because I wanted to keep it as close as possible to the upstream code from `protobuf-java`.